File: utils/file_handler.py
"""
File handling utilities
"""
import os
from pathlib import Path
import aiohttp
from utils.logger import get_logger

# Get logger
logger = get_logger(__name__)

# Create temp directory if it doesn't exist
temp_dir = Path(__file__).parent.parent / 'temp'
temp_dir.mkdir(exist_ok=True)

async def download_file(file_path: str, output_path: str) -> str:
    """
    Download a file from Telegram's servers
    
    Args:
        file_path: Telegram file path
        output_path: Path where the file should be saved
        
    Returns:
        Path to the downloaded file
    """
    try:
        # Full Telegram file URL
        url = f"https://api.telegram.org/file/bot{os.getenv('TELEGRAM_BOT_TOKEN')}/{file_path}"
        
        logger.info(f"Downloading file from {url} to {output_path}")
        
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as response:
                if response.status == 200:
                    with open(output_path, 'wb') as f:
                        f.write(await response.read())
                    logger.info(f"File downloaded successfully to {output_path}")
                    return output_path
                else:
                    raise Exception(f"Failed to download file: HTTP {response.status}")
    except Exception as e:
        logger.error(f"Error downloading file: {str(e)}", exc_info=True)
        raise Exception(f"Failed to download file: {str(e)}")

File: src/utils/fileHandler.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const { logger } = require('./logger');

// Create temp directory if it doesn't exist
const tempDir = path.join(__dirname, '../../temp');
try {
  fs.mkdirSync(tempDir, { recursive: true });
} catch (err) {
  // Directory already exists or cannot be created
}

/**
 * Download a file from a URL and save it to the temp directory
 * @param {String} url - URL of the file to download
 * @param {String} fileName - Name to save the file as
 * @returns {Promise<String>} Path to the downloaded file
 */
async function downloadFile(url, fileName) {
  try {
    // Generate a unique filename
    const uniqueFileName = `${Date.now()}-${fileName}`;
    const filePath = path.join(tempDir, uniqueFileName);
    
    logger.info(`Downloading file from ${url} to ${filePath}`);
    
    // Download the file
    const response = await axios({
      method: 'GET',
      url: url,
      responseType: 'stream',
      maxContentLength: Infinity,
      maxBodyLength: Infinity
    });
    
    // Create a write stream and pipe the response to it
    const writer = fs.createWriteStream(filePath);
    response.data.pipe(writer);
    
    // Return a promise that resolves when the file is downloaded
    return new Promise((resolve, reject) => {
      writer.on('finish', () => {
        logger.info(`File downloaded successfully to ${filePath}`);
        resolve(filePath);
      });
      writer.on('error', (err) => {
        logger.error(`Error writing file to ${filePath}:`, err);
        reject(err);
      });
    });
  } catch (error) {
    logger.error('Error downloading file:', error);
    throw new Error('Failed to download file: ' + error.message);
  }
}

module.exports = { downloadFile };